### PR TITLE
Introduce End to End testing Infrastructure

### LINKED ISSUE #1599: Implement e2e testing framework.
We need to start getting some end to end tests going for our blocks. This issue is to be focused on getting an e2e testing framework in place that we can start building e2e tests against.

- We should be able to use what has been developed for WooCommerce core as a base.
- Will need to prepare at least one set of tests to validate the tests work.
- Will need to document how to run tests locally.
- Will need to wire up to travis for automatically running e2e tests as a part of a job.

### PR DESCRIPTION
**Pull request by senadir**
This PR introduces **End-2-End** testing for our projects, it builds from WooCommerce E2E Infrastructure, and from Gutenberg Setup and Infrastructure, sharing the best of two worlds.

## What is added

1. A docker file for running tests in your machine `docker-compose.yaml` and `Dockerfile`.
2. A babel file to setup config for jest `babel.config.js`.
3. A `./tests/e2e-tests/config` folder with most configurations.
4. A `jest.setup.js` file that has some initial code that needs to be run before we run those tests.
5. woocommerce as a composer dev dependency for docker and an npm dependency for test utils.

### `jest.setup.js file`

This file contains some routine copied from GB `setup-test-framework.js` file, it mostly cleans and loads data by:

1. Capturing page events to destroy them after a test suite is done.
2. Catch console errors and directly report them.
3. Remove all existing posts and products.
4. Clear browser cache and local storage.
5. Import WooCommerce sample data.

### Prerequisites:

1. Install [Docker and docker compose](https://docs.docker.com/install/#desktop).
2. Run `composer install` so you can download WooCommerce.
3. Run `npm install` to install the new dependencies.
4. Run `npm run build`.

### Running tests:

1. inside the project, run `docker-compose up`.
	* You can run `docker-compose up -d` if you want to keep the containers running in the background.
	* The first time will take some time as it downloads the docker images.
2. In a new terminal window, run `npm run test:e2e` or `npm run test:e2e-dev` for something visual (good to understand why tests are not working). 
	* To run a specific test, either normally or in dev, run the same command and add the path to the test to it, e.g `npm run test:e2e ./tests/e2e-tests/specs/all-products/change-rows-columns.test.js`.
3. after you feel satisfied with the tests, you can stop the docker by `control+c` or simply `docker-compose down` if you have them in the background.
	* Alternatively, you can run `docker-compose down -v` to remove everything (including database entries).

## Writing tests

When writing tests, they should follow a certain pattern, they should:

1. Live in `./tests/e2e-tests/specs/`, in whatever folder you name them, we haven't decided on a pattern.
2. Be named `something.test.js`, so they should end with **test.js** in order to be picked up by jest.
3. Follow the Describe and It pattern, a `describe` is a suite and and `it` is a test.
4. While not required, tests can use snapshots for comparison, those snapshots should be committed.

Inspire from Gutenberg tests to see how tests are usually written.

You can also inspire from this PR [#1673](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1673/files).

closes #1599
## Next steps

What is left is to write the travis integration so our tests can run against each PR, after that, we can go wild writing all the e2e tests.

### CODE REVIEW

**Comment by senadir** [comment]
While writing some tests I seem to not be able to insert products, the block editor seems to be disabled for me
image

trying to investigate this issue, I've [asked around in core slack](https://wordpress.slack.com/archives/C02QB2JS7/p1580317474187000)

**Comment by senadir** [comment]
This PR is now ready to be reviewed, tested, played with, and hopefully merged, while it seems big, it’s only 651 change.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1 @@
+FROM wordpress:5.3
```
Comment (haszari) [inline Dockerfile]: If we're not running any commands to customise the container we don't need a Dockerfile, can just reference wordpress in compose file, e.g. 

```yml
  wordpress-woocommerce-blocks-dev:
     image: wordpress:5.3
```

Though as we evolve, a custom dockerfile might be useful for setting up CLI or custom permissions etc.

Is this container intended purely for e2e tests or for blocks dev generally?

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      WORDPRESS_DB_PASSWORD: wordpress
+      WORDPRESS_TABLE_PREFIX: "wp_"
+      WORDPRESS_DEBUG: 1
+    volumes:
+      - "./:/var/www/html/wp-content/plugins/woocommerce-gutenberg-products-block"
+      - "./vendor/woocommerce/woocommerce/:/var/www/html/wp-content/plugins/woocommerce"
+      - wordpress:/var/www/html
+
+  wordpress-cli:
+    depends_on:
```
Comment (haszari) [inline docker-compose.yaml]: Are we using wp-cli or is this an experiment that didn't work?

File: Dockerfile (hunk 1 above)
Comment (senadir) [reply Dockerfile]: no, for tests only, it was like that when I first started copying from wc core

File: docker-compose.yaml (hunk 2 above)
Comment (senadir) [reply docker-compose.yaml]: yes we're using it to activate plugins

File: package.json [hunk 3]
```diff
@@ -26,6 +26,8 @@
 		"release": "sh ./bin/wordpress-deploy.sh",
 		"build": "cross-env BABEL_ENV=default NODE_ENV=production webpack",
 		"build:ci": "npm run build && npm run size-check && npm run lint:js && npm run lint:css",
+		"docker:up": "docker-compose up --build -d",
+    "docker:down": "docker-compose down -v",
```
Comment (mikejolley) [inline package.json]: indentation is off here

File: Dockerfile (hunk 1 above)
Comment (haszari) [reply Dockerfile]: Aha, cool. Maybe we can move the compose setup and related config to a subfolder e.g. `e2e-tests` since it's specific to tests.

**Comment by nerrad** [comment]
I'd like to get this merged and iterate, but before we can do that we need to make sure travis builds aren't failing.  We'll also need to add documentation about clearing the `vendor` folder locally before doing builds because of the risk of including WooCommerce core pulled in via `composer` (at least until we can fix the permissions issues in the e2e test builds).

**Comment by nerrad** [comment]
closing this in favor of #1716

**Pull request closed without merging**